### ISSUE TITLE
Signal seemingly failed to restore backup, then works. _not_ due to foreign key or backup issue

### ISSUE THREAD

**Issue by sig-mag0** [signalapp/Signal-Android#12963]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

**Update**: See below comment for solution / workaround. The restore process finally finished but the app still crashed in between and the process _seemed_  to have failed.

### Bug description
I got a new Android phone (Galaxy S23) and have been trying to restore backup file(s) from my old dead Galaxy S10 phone but restoring failed without error message. On recent attempts with new Signal Beta version the backup actually seemed to finish but then Signal is stuck in an endlessly loading on the 'Verification Code' screen for >15 min. Leaving the app in any way brings me back to the intial welcome screen (same as after new install) but the app now uses 12GB of storage, which is the backup's size. I have older backups but they also failed to restore. 
There seem to be many similar reports, but I didn't find any that are quite the same and no working solution (many issues were fixed by previous updates or repairing the backup). Most similar: https://github.com/signalapp/Signal-Android/issues/12948

Using **signalbackup-tools** I tried finding if there is a problem in the backup (https://github.com/bepaald/signalbackup-tools/issues/120), but my backup seems to have **no issues**. The html export looks good and complete.
The actual backup used is from April 23, the backup I used most recently (below screenshot) states it is from today because I fiddled with the file in signalbackup-tools. Older backups show same behavior or crash immediatlz after "Restore Complete" or even during the process ("Something went wrong with Signal").

### Steps to reproduce
- Install and start Signal on new Android phone
- Start restoring backup from previous Android phone
- wait for "Restore Complete", Signal asks for phone number, verification code is automatically entered
- stuck on Verification Code screen (wait for 15+ min)
- Leaving the app > back to intial welcome screen same as after new installation of Signal
- Clear Cache and Data, try again. Same result

**Actual result:** Backup is not restored, app stuck on Verification Code screen
**Expected result:** Backup restores and Signal can be used

### Screenshots
![84390](https://github.com/signalapp/Signal-Android/assets/134394729/0786a96f-e866-4e56-b422-401e0e7ec121)
![84391](https://github.com/signalapp/Signal-Android/assets/134394729/22cb48f7-a9f3-4c3a-bede-f37cc2ecf7b5)

### Device info
**Device:** Samsung Galaxy S23 SM-S911U1
**Android version:** 13
**Signal version:** 6.19.9

### Link to debug log
https://debuglogs.org/android/6.19.9/abfc674b22e3237bb10ef3eb13f5a3fc60bac25cefa101374a2e668317cee316

**Comment by nicholas-signal** [comment]
> now app crashes on "Choose folder" screen

Do you by change have the developer option "Don't Keep Activities" turned on?

**Comment by sig-mag0** [comment]
Thank you for your reply. I don't know about that option, so I guess not? Is it a setting in signal?

**Comment by sig-mag0** [comment]
Big update: restoring the (unmodified) backup from April 23 finally worked. Again Signal crashed after finishing the restore process and I was greeted with the welcome screen on re-opening the app. However, this time I tried clicking transfer>backup again without clearing app data and cache and was asked to enter phone number (I actually picked a different backup but the restore step was skipped). The verification screen came and went smoothly and finally Signal proper opened and it seems all chats and media are restored successfully.

I had previously tried restoring again from Welcome Screen without deleting app data with no success. But maybe installing the newest beta version (6.21.2) helped.

So now the issue is not that restoring the backup fails, but that it _seemingly_ fails and the welcome screen is shown a second time with no indication that the restore actually worked.
